### ISSUE TITLE
Solution Submission error: Transaction nonce is too low.

### ISSUE THREAD

**Issue by e00E** [gnosis/dex-services#1143]
TLDR: Solution submission transaction got mined right when we submitted a new transaction with higher gas price so the new transaction failed but the solution was already correctly submitted.

We saw this error in the [logs](https://logs.gnosis.io/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:'2020-07-17T10:10:00.000Z',mode:absolute,to:'2020-07-17T10:15:00.000Z'))&_a=(columns:!(log),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'63bc0b10-5f93-11e9-bf17-fbe383fc7c9c',key:kubernetes.pod_name,negate:!f,params:(query:standard-solver-mainnet,type:phrase),type:phrase,value:standard-solver-mainnet),query:(match:(kubernetes.pod_name:(query:standard-solver-mainnet,type:phrase)))),('$state':(store:appState),meta:(alias:!n,disabled:!t,index:'63bc0b10-5f93-11e9-bf17-fbe383fc7c9c',key:log,negate:!t,params:(query:transport,type:phrase),type:phrase,value:transport),query:(match:(log:(query:transport,type:phrase)))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'63bc0b10-5f93-11e9-bf17-fbe383fc7c9c',key:log,negate:!f,params:!(driver,core),type:phrases,value:'driver,%20core'),query:(bool:(minimum_should_match:1,should:!((match_phrase:(log:driver)),(match_phrase:(log:core))))))),index:'63bc0b10-5f93-11e9-bf17-fbe383fc7c9c',interval:auto,query:(language:lucene,query:''),sort:!('@timestamp',desc))).

Relevant parts:

1. > 2020-07-17T10:10:49.899Z INFO [core::driver::stablex_driver] Computed solution for batch 531660
2. > 2020-07-17T10:10:50.104Z DEBG [core::transport] [id:3990] sending request: {"jsonrpc":"2.0","method":"eth_getTransactionCount"
3. > 2020-07-17T10:10:50.109Z DEBG [core::transport] [id:3990] received response: '{"jsonrpc":"2.0","result":"0x6ed","id":3990}'
4. > 2020-07-17T10:10:50.175Z INFO [core::solution_submission] solution submission try 0 with gas price 78100000001
5. >  2020-07-17T10:10:50.177Z DEBG [core::transport] [id:3991] sending request: '{"jsonrpc":"2.0","method":"eth_sendRawTransaction"
6. > 2020-07-17T10:12:01.179Z INFO [core::solution_submission] solution submission try 1 with gas price 90000000000
7. > 2020-07-17T10:12:01.187Z ERRO [core::driver::scheduler::system] Batch 5316601 failed with unretryable error: method 'submitSolution(uint32,uint256,address[],uint16[],uint128[],uint128[],uint16[]):(uint256)' failure: web3 error: RPC error: Error { code: ServerError(-32010), message: "Transaction nonce is too low. Try incrementing the nonce.", data: None }

In etherscan we can see that [the transaction](https://etherscan.io/tx/0x7f875963286058a93822aabbcb97828868d69e2f9e29e2bc78feae4183c5a4ab) did complete with the nonce `0x6ed == 1773` at gas price `78100000001`.

What I believe happened is that while we stopped checking for the first transaction to complete and switched to the higher gas price the first transaction did actually complete. Since that transaction completed the nonce incremented and thus we got the error about the reusing the nonce in the second attempt.
This sequence of events could already have happened in the past. Since we did not used to explicitly set the nonce we would have gotten a benign error because we would be attempting to submit the same already submitted solution again with an incremented nonce (the error would be that the solution objective value has not increased enough). The nonce would be incremented because we did not specify it, so it picks the next nonce from the node.
Now that we do keep track of the nonce we notice the error and do not classify it as benign (there is a fixed list of error strings that are benign).

One way to fix this is to also classify this nonce error as benign. This seems generally fine but could maybe cause a problem if our nonce picking logic was wrong and would keep picking an already used nonce for example. In this case we wouldn't notice the problem.
Maybe we can be smarter and try to keep track of the transactions with lower gas prices that we currently just abandon so that if we get a nonce error we can check that one of the lower gas price transactions has actually completed.

**Comment by e00E** [comment]
As Felix mentioned on Slack this could also happen when we cancel the solution submission (but it did not happen at that point in this case).

**Comment by fleupold** [comment]
It would be nice to keep track about the tx hashes of our attempts and treat cancellation failures only as benign if we see one of the tx hashed actually having a receipt with succesful execution.
